### ISSUE TITLE
let_and_return false positive

### ISSUE THREAD

**Issue by leonardo-m** [rust-lang/rust-clippy#2904]
This code:

```
fn foo<T>(spam: &[T]) -> impl Iterator<Item=(&T, &T)> {
    spam
    .iter()
    .enumerate()
    .flat_map(move |(i, x1)| spam[i ..]
                             .iter()
                             .map(move |x2| (x1, x2)))
}
fn bar() -> usize {
    let data = vec![vec![1, 2], vec![3, 4]];
    let result = foo(&data).count();
    result
}
fn main() {
    bar();
}
```

Clippy gives:

```
warning: returning the result of a let binding from a block. Consider returning the expression directly.
  --> src\main.rs:12:5
   |
12 |     result
   |     ^^^^^^
   |
   = note: #[warn(let_and_return)] on by default
note: this expression can be directly returned
  --> src\main.rs:11:18
   |
11 |     let result = foo(&data).count();
   |                  ^^^^^^^^^^^^^^^^^^
   = help: for further information visit https://rust-lang-nursery.github.io/rust-clippy/v0.0.212/index.html#let_and_return
```

But while the original code compiles, if I follow the suggestion I get a compilation error (that goes away using the NLL feature):

```
error[E0597]: `data` does not live long enough
  --> ...\main.rs:11:10
   |
11 |     foo(&data).count()
   |          ^^^^ borrowed value does not live long enough
12 | }
   | - `data` dropped here while still borrowed
   |
   = note: values in a scope are dropped in the opposite order they are created
```

**Comment by detrumi** [comment]
This can be closed, since NLL is now enabled on stable.

**Comment by oli-obk** [comment]
Only on the 2018 edition, but yea, ppl actively developing stuff and using clippy should probably move on to the new edition
